### PR TITLE
Some thoughts for a load more option.

### LINKED ISSUE #33: too many leaves
Would love a way to set a limit for the maximum number of leaves that should be rendered initially.  With the ability to 'load more' or something, nothing fancy haha.   Not sure if something like that fits into the scope here but i thought it would be a nice add, really helps if you have a lot of child nodes. Not sure how or if i could tap into doing something like this.

...i guess i could do something with the data i feed into react-infinity-menu through props... but that seems to defeat the purpose.

Any thoughts?

### PR DESCRIPTION
**Pull request by skoschik**
This is an incomplete but workable proof of concept for fixes #33 .  Added added a quantifier property to the top level component which is pushed down to a new property on branch nodes to track how many leafs should be visible.  Load more increments that value and displays the indexes in the visible range.  Should work with the search but I'm not using that feature in production.

### CODE REVIEW

**Comment by tiansijie** [comment]
sorry for the late response, I will take a look soon

File: src/infinity-menu.js [hunk 1]
```diff
@@ -31,7 +32,23 @@ export default class InfinityMenu extends React.Component {
 		event.preventDefault();
 		if (!this.state.search.isSearching || !this.state.search.searchInput.length) {
 			node.isOpen = !node.isOpen;
-			NestedObjects.set(tree, keyPath, node);
+			if (!node.isOpen || node.children && !node.maxLeaves) {
```
Comment (tiansijie) [inline src/infinity-menu.js]: why we need this check every time we click on the node?

File: src/infinity-menu.js [hunk 2]
```diff
@@ -31,7 +32,23 @@ export default class InfinityMenu extends React.Component {
 		event.preventDefault();
 		if (!this.state.search.isSearching || !this.state.search.searchInput.length) {
 			node.isOpen = !node.isOpen;
-			NestedObjects.set(tree, keyPath, node);
```
Comment (tiansijie) [inline src/infinity-menu.js]: Any reason why we delete this?

File: src/infinity-menu.js [hunk 3]
```diff
@@ -31,7 +32,23 @@ export default class InfinityMenu extends React.Component {
...
+	}
+
+	onLoadMoreClick(tree, node, keyPath, event) {
+		event.preventDefault();
+		const keyPathArray = keyPath.split('.')
+		const parentPath = Object.assign([],keyPathArray).splice(0, keyPathArray.length - 2)
+		const parentNode = _.get(this.props.tree, parentPath)
+		if (parentNode) {
+			parentNode.maxLeaves = (!parentNode.maxLeaves) ? this.props.maxLeaves : parentNode.maxLeaves + this.props.maxLeaves;
 			if (this.props.onNodeMouseClick) {
```
Comment (tiansijie) [inline src/infinity-menu.js]: do we need this part for onLoadMoreClick?

File: package.json [hunk 4]
```diff
@@ -20,12 +20,14 @@
   },
   "homepage": "https://github.com/socialtables/react-infinity-menu",
   "dependencies": {
-    "nested-objects": "0.0.2"
+    "babel-preset-es2015": "^6.16.0",
+    "babel-preset-react": "^6.16.0",
```
Comment (tiansijie) [inline package.json]: since babel is only use for example, we could put these into devDependencies

File: example/src/index.js [hunk 5]
```diff
@@ -81,6 +181,11 @@ class App extends React.Component {
 		this.setState({
 			tree: tree
 		});
+		setTimeout(() => {
```
Comment (tiansijie) [inline example/src/index.js]: why we setTimeout here?

File: package.json (hunk 4 above)
Comment (skoschik) [inline package.json]: right on.  also, sorry to gut nested objects.  i wasn't paying enough attention

File: example/src/index.js (hunk 5 above)
Comment (skoschik) [inline example/src/index.js]: pretty sure i was just experimenting with changing the max leaves.  a poor choice to leave it in there

File: src/infinity-menu.js (hunk 2 above)
Comment (skoschik) [inline src/infinity-menu.js]: i was playing around and forgot about it.  i'll get it back in as i work through my mess

File: src/infinity-menu.js (hunk 1 above)
Comment (skoschik) [inline src/infinity-menu.js]: you're right i don't think we do

File: src/infinity-menu.js [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import React from "react";
 import SearchInput from "./search-input";
 import NestedObjects from "nested-objects";
+import _ from 'lodash'
+// import NestedObjects from "nested-objects";
```
Comment (tiansijie) [inline src/infinity-menu.js]: could you remove this unused line?

**Comment by tiansijie** [comment]
two things I noticed when I run the example.
1. Load More option is not showing up at the first level 
screen shot 2016-11-02 at 4 58 01 pm

2. How we solving the search problem? Should search could also search the hiding options?

File: package.json [hunk 7]
```diff
@@ -20,12 +20,16 @@
...
   "homepage": "https://github.com/socialtables/react-infinity-menu",
   "dependencies": {
+    "lodash": "^4.16.2",
     "nested-objects": "0.0.2"
   },
   "devDependencies": {
     "babel": "^5.8.23",
     "babel-core": "^5.8.25",
-    "babel-loader": "^5.3.2",
+    "babel-loader": "^5.4.2",
```
Comment (tiansijie) [inline package.json]: if we want to upgrade the babel-loader, why not just updated to the latest version?

File: src/infinity-menu.js [hunk 8]
```diff
@@ -166,6 +197,29 @@ export default class InfinityMenu extends React.Component {
...
 				}
+			} else {
+				if (relativeIndex === filteredChildren.length - 1) {
+					if (this.props.loadMoreComponent) {
+							prevs.push(
+							<div key={itemKey}
+								onClick={this.onLoadMoreClick.bind(this, tree, curr, keyPath)}
+							>
+								{
+									this.props.loadMoreComponent//not really a component yet...
```
Comment (tiansijie) [inline src/infinity-menu.js]: why this is not component

**Comment by tiansijie** [comment]
Also could you please update the Readme as well? Thanks for all your work, it's really helping this component become better!

**Comment by skoschik** [comment]
I only implemented it on the leaves for now, that's all my use-case required ATM.   Also, i'm not using built in search so that could be flakey.  My search is API bound and I may have gutted some code that was related to that when i was cleaning up the pull request.  Honestly, not sure why i did a pull request on this as it's really just a hackish POC.

**Comment by skoschik** [comment]
Thanks man, I'm a big fan of this component so I'm happy to help if I can!

Also, I'll try to get some tests written along with the readme

**Comment by skoschik** [comment]
I'll put the babel version on 5.8.38 tomorrow.  I didn't realize the CI was preferring global babel until just now.

**Comment by tiansijie** [comment]
lgtm~ thanks for your help

**Pull request merged**